worldedit-core/src/main/java/com/sk89q/worldedit/regions/shape/ArbitraryBiomeShape.java

```diff
@@ -82,14 +78,14 @@ public ArbitraryBiomeShape(Region extent) {
      * @param defaultBaseBiome The default biome for the current column.
      * @return material to place or null to not place anything.
      */
-    protected abstract BiomeType getBiome(int x, int z, BiomeType defaultBaseBiome);
+    protected abstract BiomeType getBiome(int x, int y, int z, BiomeType defaultBaseBiome);
 
-    private BiomeType getBiomeCached(int x, int z, BiomeType baseBiome) {
-        final int index = (z - cacheOffsetZ) + (x - cacheOffsetX) * cacheSizeZ;
+    private BiomeType getBiomeCached(int x, int y, int z, BiomeType baseBiome) {
+        final int index = (y - cacheOffsetY) + (z - cacheOffsetZ) * cacheSizeY + (x - cacheOffsetX) * cacheSizeY * cacheSizeZ;
 
         final BiomeType cacheEntry = cache[index];
         if (cacheEntry == null) {// unknown, fetch material
-            final BiomeType material = getBiome(x, z, baseBiome);
+            final BiomeType material = getBiome(x, y, z, baseBiome);
             if (material == null) {
                 // outside
                 cache[index] = BiomeTypes.THE_VOID;
```

Comment:
Use of THE_VOID here is suspect, perhaps we can use `null` or something? I don't really like the fact that this is inconsistent _and_ can cause issues if used in a void world.